ml-agents-envs/mlagents_envs/rpc_utils.py

```diff
@@ -33,15 +33,25 @@ def behavior_spec_from_proto(
     :param agent_info: protobuf object.
     :return: BehaviorSpec object.
     """
-    observation_specs = []
-    for obs in agent_info.observations:
-        observation_specs.append(
-            ObservationSpec(
-                tuple(obs.shape),
-                tuple(DimensionProperty(dim) for dim in obs.dimension_properties),
-                ObservationType(obs.observation_type),
-            )
-        )
+    observation_shape = [tuple(obs.shape) for obs in agent_info.observations]
```

Comment:
I pushed @awjuliani to rewrite this in its current form for the `ObservationType` PR; I think the single loop is easier to read than multiple list comprehensions + zip (and probably fewer memory allocations and therefore faster). Can you move the logic for the dimension properties to a helper function?